app/models/spree/gateway/stripe_gateway.rb

```diff
@@ -80,45 +84,112 @@ def create_profile(payment)
+      if gateway_options.dig(:originator) && gateway_options[:originator].order_id.present?
+        merch_order = Spree::Order.find_by(id: gateway_options[:originator][:order_id])
+
+        apply_level3_data!(options, merch_order)
+      end
+
+      if (customer = creditcard.gateway_customer_profile_id)
         options[:customer] = customer
       end
-      if token_or_card_id = creditcard.gateway_payment_profile_id
+
+      if (token_or_card_id = creditcard.gateway_payment_profile_id)
         # The Stripe ActiveMerchant gateway supports passing the token directly as the creditcard parameter
         # The Stripe ActiveMerchant gateway supports passing the customer_id and credit_card id
         # https://github.com/Shopify/active_merchant/issues/770
         creditcard = token_or_card_id
       end
+
       return money, creditcard, options
     end
 
+    def apply_level3_data!(options, merch_order)
+      options[:merchant_reference] = merch_order.id
+      options[:customer_reference] = merch_order.guest_token
+
+      if merch_order.for_delivery?
+        options[:shipping_address_zip] = merch_order.shipping_address.zipcode
+        options[:shipping_from_zip] = merch_order.shipments.first.stock_location.zipcode
+        options[:shipping_amount] = (merch_order.shipments.first.cost * 100.0).to_i
+      end
+
+      options[:line_items] = []
+
+      merch_order.line_items.each do |li|
+        options[:line_items] << {
+          product_code: li.variant.id,
+          product_description: li.product.description.presence || li.product.name,
+          unit_cost: (li.price * 100.0).to_i,
+          quantity: li.quantity,
+          tax_amount: (li.additional_tax_total * 100.0).to_i,
+          discount_amount: (((merch_order.promo_total.abs.to_f * li.quantity.to_f) / merch_order.quantity.to_f) * 100.0).to_i
+        }
+      end
+
+      merch_order_total = merch_order.total.to_f
+      checksum_total = calculate_checksum_from_options(options).to_f
+
+      if checksum_total != merch_order_total
+        notify_exception_handler(
+          "Checksum failed when sending Stripe Level 3 data for Spree::Order",
+          context: {
+            merch_order_id: merch_order.id,
+            merch_order_total: merch_order_total,
+            checksum_total: checksum_total,
+            options: options
+          }
+        )
+
+        options.delete(:line_items)
+        options.delete(:merchant_reference)
+        options.delete(:customer_reference)
+
+        options.delete(:shipping_address_zip)
+        options.delete(:shipping_from_zip)
+        options.delete(:shipping_amount)
+      end
+    end
+
+    def notify_exception_handler(message, context)
+      HoneyBadger.notify(message, context)
+    end
+
+    def calculate_checksum_from_options(options)
+      line_item_amount = options[:line_items].sum { |li| ((li[:quantity] * li[:unit_cost]) - li[:discount_amount]) + li[:tax_amount] }.to_f
```

Comment:
Given the order of operations, technically none of the parentheses are necessary...but I'll grant you 1 set since I'm feeling generous.

```suggestion
      line_item_amount = options[:line_items].sum { |li| (li[:quantity] * li[:unit_cost]) - li[:discount_amount] + li[:tax_amount] }.to_f
```